chrome/android/java/src/org/chromium/chrome/browser/adblock/AdblockInitializer.java

```diff
@@ -0,0 +1,140 @@
+import org.adblockplus.libadblockplus.android.settings.AdblockHelper;
+import org.chromium.base.ContextUtils;
+import org.chromium.base.ThreadUtils;
+import org.chromium.base.library_loader.ProcessInitException;
+import org.chromium.chrome.browser.init.BrowserParts;
+import org.chromium.chrome.browser.init.ChromeBrowserInitializer;
+import org.chromium.chrome.browser.init.EmptyBrowserParts;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class AdblockInitializer implements SingleInstanceEngineProvider.EngineCreatedListener {
+
+    private final InitializationState mUnitializedState = new UnitializedState();
+    private final InitializationState mInitializedState = new InitializedState();
+    private InitializationState mInitializationState = mUnitializedState;
+    private final Set<OnAdblockInitListener> mInitListeners = new HashSet<>();
+
+    private AdblockInitializer() {
+    }
+
+    // Thread-safe lazy initialization
+    private static class LazyLoader {
+        private static final AdblockInitializer sInstance = new AdblockInitializer();
+    }
+
+    public static AdblockInitializer getInstance() {
+        return AdblockInitializer.LazyLoader.sInstance;
+    }
+
+    @Override
+    public void onAdblockEngineCreated(final AdblockEngine engine) {
+        ThreadUtils.postOnUiThread(() -> beInitialized(engine));
+    }
+
+    @UiThread
+    public void init(final ChromeBrowserInitializer initializer) {
+        mInitializationState.init(initializer);
+    }
+
+    public void addAdblockInitListener(final OnAdblockInitListener listener) {
+        mInitializationState.addAdblockInitListener(listener);
+    }
+
+    private void initAdblock() {
+        final Context context = ContextUtils.getApplicationContext();
+        final String basePath = context.getDir(
+                AdblockEngine.BASE_PATH_DIRECTORY,
+                Context.MODE_PRIVATE)
+                .getAbsolutePath();
+
+        final long ptr = AdblockBridge.getInstance().getIsolateProviderNativePtr();
+        AdblockHelper.get()
+                .init(context, basePath, true, AdblockHelper.PREFERENCE_NAME)
+                .addEngineCreatedListener(this)
+                .useV8IsolateProvider(ptr);
+
+        AdblockHelper.get().getProvider().retain(true);
+    }
+
+    @UiThread
+    private void beInitialized(final AdblockEngine engine) {
+        final long ptr = engine.getFilterEngine().getNativePtr();
+        AdblockBridge.getInstance().setFilterEngineNativePtr(ptr);
+        mInitializationState = mInitializedState;
+        for (final OnAdblockInitListener listener : mInitListeners) {
+            listener.onAdblockInit();
+        }
+        mInitListeners.clear();
+    }
+
+    public interface OnAdblockInitListener {
+        void onAdblockInit();
+    }
+
+    private abstract class InitializationState {
+
+        void init(final ChromeBrowserInitializer initializer) {
```

Comment:
looks strange. Why not make it `abstract` here?
I understand it's just because in most subclasses it just does nothing, but isn't it better to have empty method in each subclass?

Reply:
Yeah I made it with empty implementation just for the sake of not needing to redefine some methods that are sort of optional in the children classes. We have been doing like this in Orca.